lib/scholar/decomposition/pca.ex

```diff
@@ -0,0 +1,252 @@
+
+  @doc """
+  Fits a PCA for sample inputs `x`.
+
+  ## Options
+
+  #{NimbleOptions.docs(@fit_opts_schema)}
+
+  ## Returns
+
+  The function returns a struct with the following parameters:
+
+    * `:components` - Principal axes in feature space, representing the directions of maximum variance in the data.
+      Equivalently, the right singular vectors of the centered input data, parallel to its eigenvectors.
+      The components are sorted by `:explained_variance`.
+
+    * `:explained_variance` - The amount of variance explained by each of the selected components.
+      The variance estimation uses `:num_samples - 1` degrees of freedom.
+      Equal to `:num_components` largest eigenvalues of the covariance matrix of `x`.
+
+    * `:explained_variance_ratio` - Percentage of variance explained by each of the selected components.
+      If `:num_components` is not set then all components are stored and the sum of the ratios is equal to 1.0.
+
+    * `:singular_values` - The singular values corresponding to each of the selected components.
+      The singular values are equal to the 2-norms of the `:num_components` variables in the lower-dimensional space.
+
+    * `:mean` - Per-feature empirical mean, estimated from the training set.
+
+    * `:num_components` - It equals the parameter `:num_components`, or the lesser
+      value of `:num_features` and `:num_samples` if the parameter `:num_components` is `nil`.
+
+    * `:num_features` - Number of features in the training data.
+
+    * `:num_samples` - Number of samples in the training data.
+  """
+  deftransform fit(x, opts \\ []) do
+    fit_n(x, NimbleOptions.validate!(opts, @fit_opts_schema))
+  end
+
+  # TODO Add support for :num_components as a float when dynamic shapes will be implemented
+  defnp fit_n(x, opts \\ []) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError, "expected x to have rank equal to: 2, got: #{inspect(Nx.rank(x))}"
+    end
+
+    {num_samples, num_features} = Nx.shape(x)
+    num_components = opts[:num_components]
+
+    mean = Nx.mean(x, axes: [0])
+    x = x - mean
+    {decomposer, singular_values, components} = Nx.LinAlg.svd(x)
+
+    {num_components, trim_dim} =
+      calculate_num_components(
+        num_components,
+        num_features,
+        num_samples
+      )
+
+    {_, components} = flip_svd(decomposer[[0..-1//1, 0..(trim_dim - 1)]], components)
+    components = components[[0..(num_components - 1), 0..-1//1]]
+
+    explained_variance = singular_values * singular_values / (num_samples - 1)
+    explained_variance_ratio = explained_variance / Nx.sum(explained_variance)
+
+    %__MODULE__{
+      components: components,
+      explained_variance: explained_variance,
+      explained_variance_ratio: explained_variance_ratio,
+      singular_values: singular_values,
+      mean: mean,
+      num_components: num_components,
+      num_features: num_features,
+      num_samples: num_samples
+    }
```

Comment:
I think we should slice all the values to only include `:num_component`. That's what sklearn does as well.